src/helics/core/FilterFederate.cpp

```diff
@@ -0,0 +1,995 @@
+                        }
+                    }
+                } else {
+                    bool destFilter = (cmd.action() == CMD_SEND_FOR_DEST_FILTER_AND_RETURN);
+                    bool returnToSender =
+                        ((cmd.action() == CMD_SEND_FOR_FILTER_AND_RETURN) || destFilter);
+                    auto source = cmd.getSource();
+                    auto mid = cmd.messageID;
+                    auto tempMessage = createMessageFromCommand(std::move(cmd));
+                    tempMessage = FiltI->filterOp->process(std::move(tempMessage));
+                    if (tempMessage) {
+                        cmd = ActionMessage(std::move(tempMessage));
+                    } else {
+                        cmd = CMD_IGNORE;
+                    }
+
+                    if (!returnToSender) {
+                        if (cmd.action() == CMD_IGNORE) {
+                            return;
+                        }
+                        cmd.setSource(source);
+                        cmd.dest_id = parent_broker_id;
+                        cmd.dest_handle = interface_handle();
+                        mDeliverMessage(cmd);
+                    } else {
+                        cmd.setDestination(source);
+                        if (cmd.action() == CMD_IGNORE) {
+                            cmd.setAction(destFilter ? CMD_NULL_DEST_MESSAGE : CMD_NULL_MESSAGE);
+                            cmd.messageID = mid;
+                            mDeliverMessage(cmd);
+                            return;
+                        }
+                        cmd.setAction(destFilter ? CMD_DEST_FILTER_RESULT : CMD_FILTER_RESULT);
+
+                        cmd.source_handle = FiltI->handle;
+                        cmd.source_id = mFedID;
+                        mDeliverMessage(cmd);
+                    }
+                }
+            } else {
+                // the filter didn't have a function or was deactivated but still was requested to
+                // process
+                bool destFilter = (cmd.action() == CMD_SEND_FOR_DEST_FILTER_AND_RETURN);
+                bool returnToSender =
+                    ((cmd.action() == CMD_SEND_FOR_FILTER_AND_RETURN) || destFilter);
+                auto source = cmd.getSource();
+                if (!returnToSender) {
+                    cmd.setAction(CMD_SEND_MESSAGE);
+                    cmd.dest_id = parent_broker_id;
+                    cmd.dest_handle = interface_handle();
+                    mDeliverMessage(cmd);
+                } else {
+                    cmd.setDestination(source);
+                    cmd.setAction(destFilter ? CMD_DEST_FILTER_RESULT : CMD_FILTER_RESULT);
+
+                    cmd.source_handle = FiltI->handle;
+                    cmd.source_id = mFedID;
+                    mDeliverMessage(cmd);
+                }
+            }
+        } else {
+            assert(false);
+            // this is an odd condition (not sure what to do yet)
```

Comment:
Is this something that should be a TODO?

Is there a way to recover or gracefully shutdown so assert isn't needed?

Reply:
A lot of the FilterFederate code is pulled from the CommonCore so this just got copied over.  I don't think we have reached that condition so I haven't dealt with it yet.